training/prepare_dataset.py

```diff
@@ -419,192 +379,257 @@ def main():
-        prompts_usable = prompts_usable[start_index:end_index]
-        video_paths_usable = video_paths_usable[start_index:end_index]
-        videos = videos[start_index:end_index]
+        dataset.prompts = dataset.prompts[start_index:end_index]
+        dataset.video_paths = dataset.video_paths[start_index:end_index]
     else:
         pass
 
-    device = torch.device(f"cuda:{local_rank}")
+    rank_dataset_size = len(dataset)
 
-    if not args.save_tensors:
-        save_videos(videos, video_paths_usable, prompts_usable, pathlib.Path(args.output_dir), args.target_fps)
-    else:
-        dtype = DTYPE_MAPPING[args.dtype]
-        tokenizer = T5Tokenizer.from_pretrained(args.model_id, subfolder="tokenizer")
-        text_encoder = T5EncoderModel.from_pretrained(args.model_id, subfolder="text_encoder", torch_dtype=dtype)
-        text_encoder = text_encoder.to(device)
+    # 2. Dataloader
+    def collate_fn(data):
+        prompts = [x["prompt"] for x in data[0]]
 
-        prompt_embeds_list = []
+        images = [x["image"] for x in data[0]]
+        images = torch.stack(images).to(dtype=weight_dtype, non_blocking=True)
 
-        if rank == 0:
-            iterator = tqdm(range(0, len(prompts_usable), args.batch_size), desc="Encoding prompts")
-        else:
-            iterator = range(0, len(prompts_usable), args.batch_size)
+        videos = [x["video"] for x in data[0]]
+        videos = torch.stack(videos).to(dtype=weight_dtype, non_blocking=True)
 
-        for start_index in iterator:
-            end_index = min(len(prompts_usable), start_index + args.batch_size)
-            batch_prompts = prompts_usable[start_index:end_index]
+        return {
+            "images": images,
+            "videos": videos,
+            "prompts": prompts,
+        }
 
-            prompt_embeds = compute_prompt_embeddings(
-                tokenizer,
-                text_encoder,
-                batch_prompts,
-                max_sequence_length=args.max_sequence_length,
-                device=device,
-                dtype=dtype,
-            )
-            prompt_embeds_list.append(prompt_embeds.to("cpu"))
-
-        prompt_embeds = None
-        if len(prompt_embeds_list) > 0:
-            prompt_embeds = torch.cat(prompt_embeds_list)
-
-        del tokenizer, text_encoder
-        gc.collect()
-        torch.cuda.empty_cache()
-        torch.cuda.synchronize(device)
+    dataloader = DataLoader(
+        dataset,
+        batch_size=1,
+        sampler=BucketSampler(dataset, batch_size=args.batch_size, shuffle=True, drop_last=False),
+        collate_fn=collate_fn,
+        num_workers=args.dataloader_num_workers,
+        pin_memory=args.pin_memory,
+    )
 
-        vae = AutoencoderKLCogVideoX.from_pretrained(args.model_id, subfolder="vae", torch_dtype=dtype)
-        vae = vae.to(device)
+    # 3. Prepare models
+    device = f"cuda:{rank}"
+    
+    generator = torch.Generator(device).manual_seed(args.seed)
```

Comment:
Maybe better to always initialize the seed on a CPU.

Reply:
hmm, setting it to cpu is giving some weird device mismatch errors :thinking: i will try debugging soon, but for now, on a given gpu and same seed, seems to be fully reproducible (but yes i know why cpu is better alternative). todo